src/driver.ml

```diff
@@ -708,6 +708,38 @@ let load_input (kind : Kind.t) fn input_name ~relocate ic =
     | Impl -> input_name, Impl (Parse.implementation lexbuf)
 ;;
 
+let load_input_run_as_ppx ~input_fn ic =
```

Comment:
Out of curiosity, why can't we use `load_input` here? Don't they at least share a common part that we could extract?

Reply:
Ok from a further reading it seems there wouldn't be too much to share besides the call to `Ast_io.read` and the extraction of the `ast` and `kind` so it's not clear there's any real benefit here. What do you think?

Reply:
I tried to re-use `load_input` at the beginning, but there's really not much in common. 

I'm not really sure if I understand how we could extract the call to `Ast_io.read`, since that's already an independent function which we call in both `load_input` and `load_input_run_as_ppx`, right? So that function call is all the two `load_input` functions have in common and that's already extracted. Or were you meaning something different? `kind` is treated differently in the two functions: in `load_input` we check if `kind` matches the extension of the file given as input, which isn't possible in `load_input_run_as_ppx` since the file given as input is a temporary file. Also the way the `Ast_io.read` return result is handled is different: if `Ast_io.read` returns an `Error Not_a_binary_ast`, in `load_input` we want to treat the input as Ocaml code input, whereas in `load_input_run_as_ppx`, we want to return an error since there we expect a binary AST.

Reply:
I think I was suggesting we extract something like:
```ocaml
let read_ast ic =
  Ast_io.read ic >>= fun (ast_input_name, ast) ->
    let ast = Intf_or_impl.of_ast_io ast in
    let kind = Intf_or_impl.kind ast in
    Ok (ast_input_name, ast, kind))
```
but again, probably not worth it!

Reply:
Ooh, now I see. I like the idea, thanks! I've implemented it (in `Result`, we don't have the map/bind infix operators, that's why I've pattern matched).